### PR TITLE
gherkin/python: Support gherkin markdown

### PR DESCRIPTION
**Pull request by temyers**
## Summary

Ports the Javascript implementation of [Gherkin Markdown](https://github.com/cucumber/common/blob/main/gherkin/MARKDOWN_WITH_GHERKIN.md) support to Python.

## Details

This implementation ports the specifications and implementation from the Javascript for a (hopefully) like-for-like implementation.

## Motivation and Context

Markdown provides a much richer living documentation format over classic Gherkin syntax

Relates to: [1979](https://github.com/cucumber/common/issues/1979)

## How Has This Been Tested?

Unit tests have been ported from the Javascript version.
Additional unit tests have been added for improved coverage.
A few additional changes have been made without unit test coverage in order to ensure parity with the Javascript version.  
I presume these are covered under acceptance tests?  

Tested using `pytest` unit tests only at this stage.

Testing Environment: python-3.10 container.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [ ] The change has been ported to Java.
- [ ] The change has been ported to Ruby.
- [ ] The change has been ported to JavaScript.
- [ ] The change has been ported to Go.
- [ ] The change has been ported to .NET.
- [x] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated the CHANGELOG accordingly.

### CODE REVIEW

File: gherkin/python/gherkin/token_matcher.py
Comment (temyers) [inline gherkin/python/gherkin/token_matcher.py:193]: This fails to match for French (unicode) step.

Might need some help on this bit.

File: gherkin/python/test/gherkin_in_markdown_token_matcher_test.py
Comment (temyers) [inline gherkin/python/test/gherkin_in_markdown_token_matcher_test.py:24]: I have resolved test failures in python 2.x by expliticly using unicode strings within the test.

Is this suitable/valid use or is it producing a false negative?

Comment (mpkorstanje) [reply gherkin/python/test/gherkin_in_markdown_token_matcher_test.py:24]: I'm not sure. What is Pythons default encoding? What did change between 2/3?

Comment (temyers) [reply gherkin/python/test/gherkin_in_markdown_token_matcher_test.py:24]: @mpkorstanje 
String encoding was a key change between 2 & 3.

https://stackoverflow.com/questions/3547534/what-encoding-do-normal-python-strings-use

> In Python 2: Normal strings (Python 2.x str) don't have an encoding: they are raw data.
>
> In Python 3: These are called "bytes" which is an accurate description, as they are simply sequences of bytes, which can be text encoded in any encoding (several are common!) or non-textual data altogether.
>
> For representing text, you want unicode strings, not byte strings. By "unicode strings", I mean unicode instances in Python 2 and str instances in Python 3. Unicode strings are sequences of unicode codepoints represented abstractly without an encoding; this is well-suited for representing text.

https://portingguide.readthedocs.io/en/latest/strings.html

My interpretation is:
* Strings `'something'` are unicode by default in Python 3 - hence no need to explicitly mark the keyword as unicode.
* The non-ASCII character `é` is not encoded correctly without explicit unicode conversion in Python 2.

Therefore I think this resolution is valid and not a false negative.

Do you agree with my understanding?

Comment (mpkorstanje) [reply gherkin/python/test/gherkin_in_markdown_token_matcher_test.py:24]: Yeah that sounds reasonable. Thoug then I wonder what the encoding is of the source file in Python 2.

Comment (temyers) [reply gherkin/python/test/gherkin_in_markdown_token_matcher_test.py:24]: [token_scanner](https://github.com/cucumber/common/blob/dd87c6d272e41148ec61cb75ea22625963251b30/gherkin/python/gherkin/token_scanner.py#L22) reads in the file and specifies UTF-8 - so that adds further weight.

I'm happy this shows that my change is valid rather than a false negative.

Comment (mpkorstanje) [reply gherkin/python/test/gherkin_in_markdown_token_matcher_test.py:24]: No problems with the change. I'm not a python expert. 

I wouldn't even be opposed to dropping V2 support. Maybe on a separate PR thoug.

Comment (temyers) [reply gherkin/python/test/gherkin_in_markdown_token_matcher_test.py:24]: Lol, wouldn't call myself a python expert either.  

I'll avoid commenting either way regarding version support.  I can live without V2 support.

Is there anything else you think is required before this change can be merged?
I think I've covered cases equally comprehensively as the javascript implementation.

**Comment by temyers** [comment]
As an aside, it's great to see how clean the codebase is.  Very accessible for new contributors.

I was able to port the markdown implementation with no prior experience of the codebase based purely on the Typescript implementation and tests.

File: gherkin/python/test/gherkin_in_markdown_token_matcher_test.py
Comment (mpkorstanje) [reply gherkin/python/test/gherkin_in_markdown_token_matcher_test.py:24]: I'll try and have a look today. I would like to get this merged so I can split out Gherkin to it's own repository.

File: .devcontainer/devcontainer.json
Comment (mpkorstanje) [inline .devcontainer/devcontainer.json:4]: Was plain Ubuntu sufficient?

File: .vscode/settings.json
Comment (mpkorstanje) [inline .vscode/settings.json:7]: I think this file should go into `.gitignore`.

File: .devcontainer/devcontainer.json
Comment (mpkorstanje) [inline .devcontainer/devcontainer.json:20]: And maybe this file should go into `.gitignore` too.

File: gherkin/python/gherkin/token_matcher.py
Comment (mpkorstanje) [inline gherkin/python/gherkin/token_matcher.py:22]: I'm not sure about the conventions in python, should this class not be in it's own file?

**Review by mpkorstanje — CHANGES_REQUESTED** [review]
It looks like the acceptance tests were not updated to include Markdown. For example see:

https://github.com/cucumber/common/blob/dd87c6d272e41148ec61cb75ea22625963251b30/gherkin/javascript/Makefile#L3-L4

File: .devcontainer/devcontainer.json
Comment (mpkorstanje) [reply .devcontainer/devcontainer.json:4]: Ah nevermind. It used the Dockerfile. This is just the name.

Comment (temyers) [reply .devcontainer/devcontainer.json:4]: Yes, ubuntu is the base container.

File: .vscode/settings.json
Comment (temyers) [reply .vscode/settings.json:7]: This is IDE specific: provides a development container co-located with your repository:

https://code.visualstudio.com/docs/devcontainers/containers

It could provide an easier experience for contributors to get started as well as standardise your development environment.

I created these before I found the [cucumber/build](https://github.com/cucumber/build) repository that I could reference.

I'll remove.

File: gherkin/python/gherkin/token_matcher.py
Comment (temyers) [reply gherkin/python/gherkin/token_matcher.py:22]: :+1:

Will resolve.

**Comment by temyers** [comment]
@mpkorstanje - Updates made as per comments.  I've updated the acceptance tests to include the features.

Can this make the cut?  I note you said 'soon' ;)

Otherwise I'll create a new PR

**Comment by aslakhellesoy** [comment]
Hi @temyers,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Comment by mpkorstanje** [comment]
Cheers, looks good enough. I reckon any problems can be fixed in future iterations.

**Pull request merged**
